File: Baekjoon/Class3/23.2630 색종이 만들기.js
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8").trim().split("\n");

const n = Number(input[0]);
const paper = new Array();
for(let i = 1; i <= n; i++)
    paper.push(input[i].split(" ").map(Number));

const isUniform = (r, c, size) => {
    const first = paper[r][c];
    for(let i = r; i < r + size; i++)
        for(let j = c; j < c + size; j++)
            if(first !== paper[i][j])
                return false;
    return true;
}

const cutPapers = (r, c, size) => { // return [white, blue]
    if(isUniform(r, c, size)) // 모두 같은 색일 때 종료
        return paper[r][c] === 0 ? [1, 0] : [0, 1];

    const half = size >> 1; // 사등분
    let white = 0, blue = 0;
    for(let i = 0; i < 2; i++)
        for(let j = 0; j < 2; j++) {
            const [w, b] = cutPapers(r + i * half, c + j * half, half);
            white += w;
            blue += b;
        }
    return [white, blue];
}

const counts = cutPapers(0, 0, n);
console.log(counts.join("\n"));

File: Baekjoon/Class3/23.2630 색종이 만들기.py
import sys
input = sys.stdin.readline

def is_uniform(r, c, size, paper):
    first = paper[r][c]
    for i in range(r, r + size):
        for j in range(c, c + size):
            if first != paper[i][j]:
                return False
    return True

def cut_papers(r, c, size, paper): #return (white, blue)
    if is_uniform(r, c, size, paper): #모두 같은 색일 때 종료
        return (1, 0) if paper[r][c] == 0 else (0, 1)

    half = size // 2 #사등분
    white = 0
    blue = 0
    for i in range(2):
        for j in range(2):
            (w, b) = cut_papers(r + i * half, c + j * half, half, paper)
            white += w
            blue += b
    return (white, blue)

def main():
    n = int(input())
    paper = [list(map(int, input().split())) for _ in range(n)]
    
    (w, b) = cut_papers(0, 0, n, paper)
    print(f"{w}\n{b}")

main()